### PR TITLE
1611 interruptable ssh exploit

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Adds interrupts to SSH exploit. Related #1611 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] ~Is the TravisCI build passing?~
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running SSH exploit and checking timeouts locally.
* [x] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

image

### CODE REVIEW

File: monkey/infection_monkey/exploit/sshexec.py [hunk 1]
```diff
@@ -14,9 +13,16 @@
 from infection_monkey.telemetry.attack.t1222_telem import T1222Telem
 from infection_monkey.utils.brute_force import generate_identity_secret_pairs
 from infection_monkey.utils.commands import build_monkey_commandline
+from infection_monkey.utils.threading import interruptable_iter
+from infection_monkey.utils.timer import Timer
 
 logger = logging.getLogger(__name__)
 SSH_PORT = 22
+SSH_CONNECT_TIMEOUT = 10
```
Comment (VakarisZ) [inline monkey/infection_monkey/exploit/sshexec.py:21]: What about using timeouts from monkey/common/common_consts/timeouts.py?

File: monkey/infection_monkey/exploit/sshexec.py [hunk 2]
```diff
@@ -156,7 +193,9 @@ def _exploit_host(self) -> ExploiterResultData:
 
         try:
             with ssh.open_sftp() as ftp:
```
Comment (VakarisZ) [inline monkey/infection_monkey/exploit/sshexec.py:195]: The default tiemout of this is 1h. This might be a problem

Comment (VakarisZ) [reply monkey/infection_monkey/exploit/sshexec.py:195]: Can you do a test? Set a breakpoint at this line, kill the victim and step over. See how long it takes for this line to execute

**Review by VakarisZ — CHANGES_REQUESTED** [review]
Need you to test one thing, see comments

File: monkey/infection_monkey/exploit/sshexec.py (hunk 2 above)
Comment (ilija-lazoroski) [reply monkey/infection_monkey/exploit/sshexec.py:195]: open_sftp doesn't seem to have timeout .... 

image

Comment (ilija-lazoroski) [reply monkey/infection_monkey/exploit/sshexec.py:195]: Actually you are right. If a machine gets killed after we connect, `open_sftp` will wait one hour for a connection and block.

**Review by VakarisZ — APPROVED** [review]
Good job

File: monkey/infection_monkey/exploit/sshexec.py [hunk 3]
```diff
@@ -73,15 +97,31 @@ def exploit_with_login_creds(self, port) -> paramiko.SSHClient:
...
+                    username=user,
+                    password=current_password,
+                    port=port,
+                    timeout=SSH_CONNECT_TIMEOUT,
+                    auth_timeout=SSH_AUTH_TIMEOUT,
+                    banner_timeout=SSH_BANNER_TIMEOUT,
+                )
 
                 logger.debug("Successfully logged in %r using SSH. User: %s", self.host, user)
                 self.add_vuln_port(port)
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/sshexec.py:123]: Do we still need this call? If so, it's missing from `exploit_with_ssh_keys()`

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/exploit/sshexec.py (hunk 3 above)
Comment (ilija-lazoroski) [reply monkey/infection_monkey/exploit/sshexec.py:123]: It is used in smb, wmi and msql. I will add it in `exploit with ssh_keys()`.

File: monkey/infection_monkey/exploit/sshexec.py [hunk 4]
```diff
@@ -114,16 +155,18 @@ def _exploit_host(self) -> ExploiterResultData:
...
                 logger.error(self.exploit_result.error_message)
                 return self.exploit_result
 
+        if self._is_interrupted():
+            self._set_interrupted()
+            return self.exploit_result
+
         if not self.host.os.get("type"):
             try:
                 _, stdout, _ = ssh.exec_command("uname -o")
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/sshexec.py:172]: Missing timeout

**Review by mssalvatore — CHANGES_REQUESTED** [review]

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**